plugin/evm/validators/state_test.go

```diff
@@ -0,0 +1,286 @@
+
+	// delete
+	require.NoError(state.DeleteValidator(vID))
+
+	// write state, should reflect to DB
+	require.NoError(state.WriteState())
+	require.False(db.Has(vID[:]))
+}
+
+func TestParseValidator(t *testing.T) {
+	testNodeID, err := ids.NodeIDFromString("NodeID-CaBYJ9kzHvrQFiYWowMkJGAQKGMJqZoat")
+	require.NoError(t, err)
+	type test struct {
+		name        string
+		bytes       []byte
+		expected    *validatorData
+		expectedErr error
+	}
+	tests := []test{
+		{
+			name:  "nil",
+			bytes: nil,
+			expected: &validatorData{
+				LastUpdated: 0,
+				StartTime:   0,
+			},
+			expectedErr: nil,
+		},
+		{
+			name:  "empty",
+			bytes: []byte{},
+			expected: &validatorData{
+				LastUpdated: 0,
+				StartTime:   0,
+			},
+			expectedErr: nil,
+		},
+		{
+			name: "valid",
+			bytes: []byte{
+				// codec version
+				0x00, 0x00,
+				// up duration
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x5B, 0x8D, 0x80,
+				// last updated
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x0D, 0xBB, 0xA0,
+				// node ID
+				0x7e, 0xef, 0xe8, 0x8a, 0x45, 0xfb, 0x7a, 0xc4,
+				0xb0, 0x59, 0xc9, 0x33, 0x71, 0x0a, 0x57, 0x33,
+				0xff, 0x9f, 0x4b, 0xab,
+				// start time
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x5B, 0x8D, 0x80,
+				// status
+				0x01,
+			},
+			expected: &validatorData{
+				UpDuration:  time.Duration(6000000),
+				LastUpdated: 900000,
+				NodeID:      testNodeID,
+				StartTime:   6000000,
+				IsActive:    true,
+			},
+		},
+		{
+			name: "invalid codec version",
+			bytes: []byte{
+				// codec version
+				0x00, 0x02,
+				// up duration
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x5B, 0x8D, 0x80,
+				// last updated
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x0D, 0xBB, 0xA0,
+			},
+			expected:    nil,
+			expectedErr: codec.ErrUnknownVersion,
+		},
+		{
+			name: "short byte len",
+			bytes: []byte{
+				// codec version
+				0x00, 0x00,
+				// up duration
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x5B, 0x8D, 0x80,
+				// last updated
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x0D, 0xBB, 0xA0,
+			},
+			expected:    nil,
+			expectedErr: wrappers.ErrInsufficientLength,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			var data validatorData
+			err := parseValidatorData(tt.bytes, &data)
+			require.ErrorIs(err, tt.expectedErr)
+			if tt.expectedErr != nil {
+				return
+			}
+			require.Equal(tt.expected, &data)
+		})
+	}
+}
+
+func TestStateListener(t *testing.T) {
+	require := require.New(t)
+	db := memdb.New()
+	state, err := NewState(db)
+	require.NoError(err)
+
+	expectedvID := ids.GenerateTestID()
+	expectedNodeID := ids.GenerateTestNodeID()
+	expectedStartTime := time.Now()
+
+	// add listener
+	listener := &testCallbackListener{
+		t: t,
+		onAdd: func(vID ids.ID, nodeID ids.NodeID, startTime uint64, isActive bool) {
+			require.Equal(expectedvID, vID)
+			require.Equal(expectedNodeID, nodeID)
+			require.Equal(uint64(expectedStartTime.Unix()), startTime)
+			require.True(isActive)
```

Comment:
should probably have some verification that these callbacks are actually invoked

Reply:
dunno what I was thinking, mocks are much easier...